### PR TITLE
ctl: fix https client panic

### LINKED ISSUE #8237: panic when exec "pd health" with tls
### What did you do?
1、tiup deploy cluster
2、exec "pd health" with tls
`tiup ctl:nightly pd -u https://pd1-peer:2379 health --cacert /root/.tiup/storage/cluster/clusters/tidbcluster/tls/ca.crt --cert /root/.tiup/storage/cluster/clusters/tidbcluster/tls/client.crt --key /root/.tiup/storage/cluster/clusters/tidbcluster/tls/client.pem
Starting component ctl: /root/.tiup/components/ctl/v8.2.0-alpha-nightly/ctl pd -u https://pd1-peer:2379 health --cacert /root/.tiup/storage/cluster/clusters/tidbcluster/tls/ca.crt --cert /root/.tiup/storage/cluster/clusters/tidbcluster/tls/client.crt --key /root/.tiup/storage/cluster/clusters/tidbcluster/tls/client.pem`

### What did you expect to see?
cmd can success

### What did you see instead?
cmd panic:

`panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x11afa05]

goroutine 1 [running]:
github.com/tikv/pd/tools/pd-ctl/pdctl/command.showHealthCommandFunc(0xc000599800, {0x14c4cf6?, 0x4?, 0x14c4bae?})
        /workspace/source/pd/tools/pd-ctl/pdctl/command/health_command.go:33 +0x25
github.com/spf13/cobra.(*Command).execute(0xc000599800, {0xc00034af00, 0x8, 0x8})
        /root/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0xaa3
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000bb800)
        /root/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        /root/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/tikv/pd/tools/pd-ctl/pdctl.MainStart({0xc00012c0b0, 0x9, 0x9})
        /workspace/source/pd/tools/pd-ctl/pdctl/ctl.go:105 +0x228
main.main()
        /workspace/source/pd/tools/pd-ctl/main.go:67 +0x3cf`

### What version of PD are you using (`pd-server -V`)?
    "binary_version": "v8.2.0-alpha-63-g19c9852",
    "git_hash": "19c9852decda4cb49a2319b453c4f01c6a26014f"

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: Close #8237, Ref #7300

### What is changed and how does it work?

```commit-message
fix panic when call pd-ctl cluster with tls
```

### Check List

Tests

- Unit test

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: tools/pd-ctl/pdctl/command/cluster_command.go [hunk 1]
```diff
@@ -19,10 +19,10 @@ import "github.com/spf13/cobra"
 // NewClusterCommand return a cluster subcommand of rootCmd
 func NewClusterCommand() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:               "cluster",
-		Short:             "show the cluster information",
-		PersistentPreRunE: requirePDClient,
-		Run:               showClusterCommandFunc,
+		Use:     "cluster",
+		Short:   "show the cluster information",
+		PreRunE: requirePDClient,
```
Comment (okJiang) [inline tools/pd-ctl/pdctl/command/cluster_command.go:24]: https://github.com/tikv/pd/blob/199b01792159e5d8e83ef419a5053401e998bb0e/tools/pd-ctl/pdctl/ctl.go#L40

`RequireHTTPSClient` should be called firstly

Comment (rleungx) [reply tools/pd-ctl/pdctl/command/cluster_command.go:24]: PersistentPreRunE is called before PreRunE?

Comment (JmPotato) [reply tools/pd-ctl/pdctl/command/cluster_command.go:24]: Why change to `PreRunE`? Using `PersistentPreRunE` ensures that all sub-commands can inherit this requirement without setting `PreRunE` for each.

Comment (okJiang) [reply tools/pd-ctl/pdctl/command/cluster_command.go:24]: > PersistentPreRunE is called before PreRunE?

Yes.

> Why change to PreRunE? Using PersistentPreRunE ensures that all sub-commands can inherit this requirement without setting PreRunE for each.

For this purpose, how about we put `command.RequireHTTPSClient` into function `requirePDClient`? @JmPotato

Comment (JmPotato) [reply tools/pd-ctl/pdctl/command/cluster_command.go:24]: LGTM.

File: tools/pd-ctl/pdctl/command/global.go [hunk 2]
```diff
@@ -55,23 +55,20 @@ var PDCli pd.Client
...
-		keyPath, err = cmd.Flags().GetString("key")
-		if err != nil {
+	tlsConfig, err = parseTLSConfig(cmd)
+	if err != nil {
+		return err
+	}
+	if tlsConfig != nil {
+		if err = initHTTPSClient(tlsConfig); err != nil {
 			return err
 		}
```
Comment (JmPotato) [inline tools/pd-ctl/pdctl/command/global.go:68]:
```suggestion
		return initHTTPSClient(tlsConfig)
```

File: tools/pd-ctl/pdctl/command/global_test.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	"testing"
+
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseTLSConfig(t *testing.T) {
+	re := require.New(t)
+
+	rootCmd := &cobra.Command{
```
Comment (JmPotato) [inline tools/pd-ctl/pdctl/command/global_test.go:28]: I tend to only introduce unit tests for functions like `RequireHTTPSClient`/`requirePDClient`/`parseTLSConfig` in the `global_test.go` file. Introducing cobra commands seems a bit strange here.

Comment (okJiang) [reply tools/pd-ctl/pdctl/command/global_test.go:28]: the input of these function are `cmd *cobra.Command`. If do not introduce it, how do we test it?

Comment (okJiang) [reply tools/pd-ctl/pdctl/command/global_test.go:28]: Or do you mean introducing a new Command?

File: tools/pd-ctl/pdctl/command/global.go (hunk 2 above)
Comment (okJiang) [reply tools/pd-ctl/pdctl/command/global.go:68]: it need to execute `initNewPDClient` yet

File: tools/pd-ctl/pdctl/command/global_test.go (hunk 3 above)
Comment (JmPotato) [reply tools/pd-ctl/pdctl/command/global_test.go:28]: Oops, I didn't notice that.

File: tools/pd-ctl/pdctl/command/global.go [hunk 4]
```diff
@@ -111,46 +103,47 @@ func initNewPDClient(cmd *cobra.Command, opts ...pd.ClientOption) error {
...
-	tlsConfig, err := initTLSConfig(caPath, certPath, keyPath)
+// RequireHTTPSClient creates a HTTPS client if the related flags are set
+func RequireHTTPSClient(cmd *cobra.Command, _ []string) error {
+	tlsConfig, err := parseTLSConfig(cmd)
+	if err != nil || tlsConfig == nil {
+		return err
+	}
+	err = initHTTPSClient(tlsConfig)
 	if err != nil {
+		cmd.Println(err)
```
Comment (JmPotato) [inline tools/pd-ctl/pdctl/command/global.go:140]: This line appears unnecessary since `initHTTPSClient` cannot return an error.

**Review by JmPotato — APPROVED** [review]

File: tools/pd-ctl/pdctl/command/cluster_command.go (hunk 1 above)
Comment (okJiang) [reply tools/pd-ctl/pdctl/command/cluster_command.go]: After [EnableTraverseRunHooks](https://github.com/tikv/pd/pull/8239/commits/44688fc5c17554d8e9c568bccacab362816d6f1c), the execution order of `PersistentPreRunE` is `rootCmd -> subCmd`, so the problem has been solved.

**Review by rleungx — APPROVED** [review]

**Comment by JmPotato** [comment]
/check-dco |  
-- | --


<br class="Apple-interchange-newline">

**Pull request merged**